keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,775 @@
+, backoffLimit, and [backoffLimitPerIndex](https://kubernetes.io/docs/concepts/workloads/controllers/job/#backoff-limit-per-index)
+in the following situations:
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
+- when the job meets the successPolicy and the number of failed pods exceeds `backoffLimit`.
+- when the job meets the successPolicy and the number of failed pods per indexes exceeds `backoffLimitPerIndex` in all indexes.
+
+To avoid the above conflicts, terminating policies are evaluated the first before successPolicies.
+This means that the terminating policies are respected rather than the successPolicies.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- Test cases:
+  - tests for Defaulting and Validating
+  - verify whether a job has a SuccessCriteriaMet condition if the job meets to successPolicy and some indexes fail.
+  - verify whether a job has both complete and SuccessCriteriaMet conditions 
+  if the job meets to successPolicy and all pods are terminated
+  - verify whether a job has a failed condition if the job meets to both successPolicy and terminating policies
+  in the same reconcile cycle
+
+- `k8s.io/kubernetes/pkg/controller/job`: `2 February 2024` - `91.5%`
+- `k8s.io/kubernetes/pkg/apis/batch/validation`: `2 February 2024` - `98.0%`
+
+##### Integration tests
+
+- Test scenarios:
+  - enabling, disabling and re-enabling of the `JobSuccessPolicy` feature gate
+  - handling of successPolicy when all indexes succeeded
+  - handling of the `.spec.successPolicy.criteria.succeededIndexes` when some indexes remain pending
+  - handling of the `.spec.successPolicy.criteria.succeededCount` when some indexes remain pending 
+  - handling of successPolicy when some indexes of job with `backOffLimitPerIndex` fail
+
+##### e2e tests
+
+- Test scenarios:
+  - handling of successPolicy when all indexes succeeded
+  - handling of the `.spec.successPolicy.criteria.succeededIndexes` when some indexes remain pending
+  - handling of the `.spec.successPolicy.criteria.succeededCount` when some indexes remain pending
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind the `JobSuccessPolicy` feature gate.
+- Unit and integration tests passed as designed in [TestPlan](#test-plan).
+
+#### Beta
+
+- E2E tests passed as designed in [TestPlan](#test-plan).
+- Added a new reasons to the existing `job_finished_total` metric in [Monitoring Requirements](#monitoring-requirements).
+- Feature is enabled by default.
+- Address all issues reported by users.
+- Decided whether we introduce a `whenCriterionAchived=continue` and `whenCriterionAchived=continueWithRecreations`.
```

Comment:
This behavior has significant repercussion for both this immediate controller and fanout from there (cronjobs in tree and other consumers like UIs out of tree).  If `whenCriterionAchieved` comes back, we'll need a second alpha.

Reply:
SGTM. I will mention about the second alpha stage here.

Reply:
I added a note to "Possibility for the rest of running pods to continue running after the job meets the successPolicy".

Reply:
As long as that is a choice / optional - for some use cases it defeats the purpose of having the success policy, where it's understood the other pods will continue (but you don't want them to when the specific pods of interest are complete).

Reply:
@vsoch I still believe that `whenTargetAchieved` is valuable, although there are some unresolved issues regarding the backward compatibility and compatibility for CronJob. Then, we don't have enough time to discuss the compatibility issues in the release v1.30 cycle since the KEP deadline will be this Friday.

So, I left this section "unresolved," and I believe that we can discuss the possibility of the "whenCriterionAcheived=continue" in the next release cycle (v1.31).

Reply:
I agree with David here, the addition of that new field should happen before we transition to beta. Iow. one more alpha with additional field, and only in the next release transition to beta.

Reply:
> I agree with David here, the addition of that new field should happen before we transition to beta. Iow. one more alpha with additional field, and only in the next release transition to beta.

That makes sense.

Reply:
I added the optional second alpha stage.